In [1]:
import pandas as pd
import numpy as np
import matplotlib as mp
import math as math
import matplotlib.pyplot as plt

# Define Equation

In [2]:
def f(x):
    return np.exp(x**5)


# Midpoint Method

In [3]:
def midpoint_method(a, b, n):

    # get the delta x in the midpoint equation, the size of the interval along the boudns we are integrating
    delta_x = (b - a) / n
    # using np.linspace, get the x points for the middle of the intervals
    midpoints = np.linspace(a + delta_x/2, b - delta_x/2, n)
    #return the equation calculating the mid points
    return delta_x * np.sum(f(midpoints))

# Simpson 1/3

In [4]:
def simp13(a, b, n):
    #calculate delta_x given bounds and number of intervals
    delta_x = (b - a) / (n - 1)
    #gather x points using np.linspace
    x = np.linspace(a, b, n)
    #generate y values 
    y = f(x)

    # then find the sums of the odd x values and the sums of the even x values under the function we are integrating
    odd_sum = 0
    even_sum = 0
    for i in range(1, n-1, 2):
        odd_sum += y[i]
    for i in range(2, n-2, 2):
        even_sum += y[i]
    
    # plug it all into the equation
    return (delta_x/3) * (y[0] + 4*odd_sum + 2*even_sum + y[-1])


# Simpson 3/8

In [5]:
def simp38(a, b, n):

    #calculate the delta x given the # of intervals
    delta_x = (b - a) / (n - 1)
    #generate x values using np.linspace
    x = np.linspace(a, b, n)
    #calculate the function values we are going to integrate
    y = f(x)

    # get sums of ddifferent parts of the equation
    sum_t = 0
    sum_nt = 0
    for i in range(1, n-1):
        if i % 3 == 0:
            sum_nt += y[i]
        else:
            sum_t += y[i]
    
    # plut it all into the equation and then return 
    return (3*delta_x/8) * (y[0] + 3*sum_t + 2*sum_nt + y[-1])

# Guassian Quadrature

In [6]:
# bounds not needed since by default uses -1 to 1b
def gaussian_quadrature(n):
    #.legendre returns the xis as well as the weights (c) needed to use the gaussian quardrature formula
    xi, c = np.polynomial.legendre.leggauss(n)

    #return sum as specified in the gfaussian quadrature equation
    final = 0
    for i in range(n):
        final += c[i] * f(xi[i])

    #return
    return final


# Monte Carlo

In [7]:
def monte_carlo(a, b, n):
    # get random x values using np.uniform between the bounds and n of them
    xn = np.random.uniform(a, b, n)
    # sum of f(xn)
    sum = 0
    for i in range(n):
        sum += f(xn[i])

    # return as specified from monte carlo equation
    return ((b - a) / n) * sum

# Run and test

In [8]:
a = -1
b = 1

midpoint_res = midpoint_method(a, b, 100)
simp13_res = simp13(a, b, 101)
simp38_res = simp38(a, b, 101)
gaussian_res = gaussian_quadrature(5)
mc_res = monte_carlo(a, b, 1000)

correct_answer = 2.0949681713212


print(f"Midpoint method returns: {midpoint_res}, with an error of {midpoint_res - correct_answer}")
print(f"Simpson 1/3 method returns: {simp13_res}, with an error of {simp13_res - correct_answer}")
print(f"Simpson 3/8 method returns: {simp38_res}, with an error of {simp38_res - correct_answer}")
print(f"Gaussuan Quad method returns: {gaussian_res}, with an error of {gaussian_res - correct_answer}")
print(f"Monte Carlo method returns: {mc_res}, with an error of {mc_res - correct_answer}")

Midpoint method returns: 2.094772568546673, with an error of -0.000195602774526904
Simpson 1/3 method returns: 2.0949693708997557, with an error of 1.1995785555996008e-06
Simpson 3/8 method returns: 2.0820747975397573, with an error of -0.01289337378144273
Gaussuan Quad method returns: 2.092230298505938, with an error of -0.0027378728152620546
Monte Carlo method returns: 2.0972584667360894, with an error of 0.002290295414889343


Midpoint method returns: 2.094772568546673, with an error of -0.000195602774526904

Simpson 1/3 method returns: 2.0949693708997557, with an error of 1.1995785555996008e-06

Simpson 3/8 method returns: 2.0820747975397573, with an error of -0.01289337378144273

Gaussuan Quad method returns: 2.092230298505938, with an error of -0.0027378728152620546

Monte Carlo method returns: 2.0972584667360894, with an error of 0.002290295414889343
